wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileEnvironmentFactory.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static io.fabric8.kubernetes.client.utils.Serialization.asYaml;
+import static io.fabric8.kubernetes.client.utils.Serialization.unmarshal;
+import static java.lang.String.format;
+import static java.util.Collections.emptyMap;
+import static org.eclipse.che.api.devfile.server.Constants.KUBERNETES_TOOL_TYPE;
+import static org.eclipse.che.api.devfile.server.Constants.OPENSHIFT_TOOL_TYPE;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.KubernetesList;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import java.util.List;
+import java.util.Optional;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.devfile.model.Tool;
+import org.eclipse.che.api.workspace.server.model.impl.EnvironmentImpl;
+import org.eclipse.che.api.workspace.server.model.impl.RecipeImpl;
+
+/**
+ * Creates {@link EnvironmentImpl} from specific tool in devfile if any.
+ *
+ * @author Max Shaposhnyk
+ */
+@Singleton
+public class DevfileEnvironmentFactory {
+
+  static final String DEFAULT_RECIPE_CONTENT_TYPE = "application/x-yaml";
+
+  /**
+   * Consumes an recipe-type tool (openshift or kubernetes) from devfile and tries to create {@link
+   * EnvironmentImpl} from it (including filtering of list items using selectors, if necessary). An
+   * {@link RecipeFileContentProvider} MUST be provided in order to fetch recipe content.
+   *
+   * @param recipeTool the recipe-type tool
+   * @param recipeFileContentProvider service-specific provider of recipe file content
+   * @return optional of constructed environment from recipe type tool
+   * @throws BadRequestException when there is no content provider for recipe-type tool
+   * @throws BadRequestException when recipe-type tool content is unreachable or empty
+   */
+  public Optional<EnvironmentImpl> createEnvironment(
```

Comment:
Why we use Optional as a return value? As I can see you can get a valid EnvironmentImpl or exception.

Reply:
Yes it's reduntant